mlflow/tracking/client.py

```diff
@@ -1010,6 +1011,67 @@ def log_dict(self, run_id, dictionary, artifact_file):
                 else:
                     json.dump(dictionary, f, indent=2)
 
+    @experimental
+    def log_figure(self, run_id, figure, artifact_file):
+        """
+        Log a figure as an artifact. The following figure objects are supported:
+
+        - `matplotlib.figure.Figure`_
```

Comment:
With the current regex approach ("Option 1"): 
(+) We'll be robust to `matplotlib` or `plotly` moving around the module namespace of their `Figure` classes. In general, such API changes seem unlikely given that `plotly` and `matplotlib` both provide stable APIs.
(-) We won't detect objects that subclass matplotlib.figure.Figure - I'm not sure how common subclassing is but this seems like a potentially significant con
(-) We may have false positives with e.g. `Figure` classes in internal plotly/matplotlib modules, but it seems very unlikely that users would supply such internal classes, nor should they expect doing so to work.
(-) The code is a bit complex, but not prohibitively complex IMO

As an alternative  ("Option 2") we could check whether `plotly` / `matplotlib` are in `sys.modules`, and if so run `import plotly` in order to perform an `isinstance` check on the passed-in `figure`. In that case:
(+) The code is clearer than in the regex case - we declaratively perform a type-check instead of indirectly doing it via regex
(+) We support subclasses of `plotly` / `matplotlib` Figure
(-) The API may break against e.g. new plotly versions with this option, but not with option 1, in the case where plotly moves the module containing `Figure`.

Given these tradeoffs I think option 2 is preferable (primarily because it supports subclasses of `Figure`). If subclassing `Figure` is super uncommon then option 2 becomes less preferable, but is still better than option 1 IMO for code clarity. Lmk what you think / if I'm missing something!

Reply:
sounds good. I'll make the change.